Readme.md
![Apache 2.0 License](https://img.shields.io/badge/License-Apache%202.0-blue.svg)
![](https://travis-ci.org/goccmack/gogll.svg?branch=master)

Copyright 2019 Marius Ackerman. 
Copyright 2021 Aaron Moss.

# PEGLL
PEGLL is a fork of Marius Ackerman's GoGLL built as part of a research project to extend the GLL algorithm to handle the lookahead and ordered choice operators of Parsing Expression Grammars (PEGs).

Gogll generates a GLL or LR(1) parser and FSA-based lexer for any context-free grammar. 
The generated code is Go or Rust.

[Click here](https://goccmack.github.io/posts/2020-05-31_gogll/) for an introduction
to GLL.

See the [LR(1) documentation](doc/lr1/Readme.md) for generating LR(1) parsers.

The generated GLL parser is a clustered nonterminal parser (CNP) following 
[[Scott et al 2019](#Scott-et-al-2019)]. 
CNP is a version of generalised LL parsing (GLL) 
[[Scott & Johnstone 2016](#Scott-et-al-2016)]. 
GLL parsers can parse all context free (CF) languages.

The generated LR(1) parser is a Pager's PGM or Knuth's original LR(1) 
machine [[Pager 1977](Pager-1977)].

The generated lexer is a linear-time finite state automaton FSA 
[[Grune et al 2012](#Grune-et-al-2012)].
The lexer ignores whitespace.

Gogll accepts grammars in markdown files, which is very useful for documenting the grammar.
For example: see
[gogll's own grammar](gogll.md).

GLL has worst-case cubic time and space complexity but linear complexity for all 
LL productions [[Scott et al 2016](#Scott-et-al-2016)]. 
[See here](https://goccmack.github.io/posts/2020-05-31_gogll/) for space and
CPU time measurements of an extreme ambiguous example.
For comparable grammars tested so far gogll produces faster lexers and parsers 
than [gocc](https://github.com/goccmack/gocc) (FSA/LR-1).

# News
## 2020-08-12
From v3.2.0 gogll supports tokens that can be suppressed by the lexer. This is useful, for example, to implement code comments. See [example](examples/comments/comments.md).

## 2020-06-28
1. Gogll now also generates LR(1) parsers. It supports 
_Pager's Practical General Method, weak compatibility_ as well as 
_Knuths original LR(1) machine_ for
comparison. Pager's PGM generates LR(1) parser tables similar is size to LALR. 
The option to generate a Knuth LR(1) machine is provide for reference.  
See [LR(1) documentation](doc/lr1/Readme.md) for details.

2. Please note that the `-t <target>` option has been replace by `-go` and `-rust`.
See see [usage](#Usage) below.

## 2020-06-01
[See](https://goccmack.github.io/posts/2020-05-31_gogll/) for an introduction
to GLL and a performance comparison of the generated Go and Rust code parsers.

## 2020-05-22
GoGLL v3.1 generates Rust as well as Go parsers with similar performance:

|| Lexer | Parser | Build
|---|---|---|---|
Go | 119 μs | 1324 μs | 0.124s
Rust | 71 μs | 1297 μs | 2.932s

1. The duration was averaged over 1000 repetitions.
1. Build time was measures with the time command.
    1. For Rust: `time cargo build --release`
    2. For Go: `time go build`

See [examples/rust](examples/rust/Readme.md) for the Rust and Go programs used 
for this comparison.

Use gogll's target option to generate a Rust lexer/parser: `-t rust` (see [usage](#Usage) below). 
Gogll generates Go code by default.


## 2020-04-24
1. GoGLL now generates a linear-time FSA lexer matching the CNP parser.
1. This version of *GoGLL is faster than gocc*. It compiles a sample grammar in  
0.074 s, which GoCC compiles in 0.118 s. Gogll compiles itself in 0.041s.

# 

# Benefits and disadvantages of GLL and LR(1)
GLL is a parsing technique that can handle any context-free (CF) language. GLL has
worst case cubic time and space complexity.

LR(1) handles a subset of the context-free languages that can be parsed bottom-up
with one token look-ahead. LR(1) has linear time complexity and its table driven
parser is very efficient. Pager's _Practical General Method_ (PGM) combines
compatible states as they are generated, keeping the state space small.

A GLL parser has more expensive bookkeeping than an LR(1) parser, making the 
LR(1) parser more efficient for parsing very large inputs.

## When to use GLL
1. When the CF grammar that best expresses the problem is not LR(1).
2. When the LR(1) parser has more than a few conflicts that require additional
language symbols or complex grammar refactorisation to resolve.
3. The inputs to be parsed are not too big. GLL works very well for DSLs or 
programming languages.

## When to use LR(1)
1. When the language can be expressed as an LR(1) grammar. A grammar is LR(1) if 
gogll can generate a conflict-free parser for it.
2. When the input is very big, for example: log files containing tens of thousands
of lines.

# Motivation for a separate lexer
The following observations were made while using GoGLLv2 on a couple of projects.

* Most of the ambiguity in grammars were generated by the lexical rules.
* Handling token separation explicitly produces messy, hard to maintain grammars.
* Most of a grammar input file is whitespace, which together with the additional 
ambiguity introduced by the lexical rules, causes most of the parse time in a 
scannerless parser.
* Writing good markdown with the grammar produced slow compilations.

# Input Symbols, Markdown Files
Gogll and lexers generated by gogll accept UTF-8 input strings, which may be in 
a markdown file or a plain text file.

If the input is a markdown file gogll and lexers generated by gogll treat all 
text outside markdown code blocks as whitespace. Markdown code blocks are 
delimited by triple backticks. See [gogll.md](gogll.md) for an example.

# Gogll Grammar
Gogll v3 has a BNF grammar. See [gogll.md](gogll.md)

# Installation
1. Install Go from [https://golang.org](https://golang.org)
1. `go get github.com/goccmack/gogll` or 
1. Clone this repository and run `go install` in the root of the directory where
it is installed.

# Usage
Enter `gogll -h` or `gogll` for the following help:

```
use: gogll -h
    for help, or

use: gogll -version
    to display the version of goggl, or

use: gogll [-a][-v] [-CPUProf] [-o <out dir>] [-go] [-rust] [-gll] [-pager] [-knuth] [-resolve_conflicts] <source file>
    to generate a lexer and parser.

    <source file>: Mandatory. Name of the source file to be processed. 
        If the file extension is ".md" the bnf is extracted from markdown code 
        segments enclosed in triple backticks.
    
    -a: Optional. Regenerate all files.
        WARNING: This may destroy user editing in the LR(1) AST.
        Default: false
         
    -v: Optional. Produce verbose output, including first and follow sets,
        LR(1) sets and lexer FSA sets.
    
    -o <out dir>: Optional. The directory to which code will be generated.
                  Default: the same directory as <source file>.
                  
    -go: Optional. Generate Go code.
          Default: true, but false if -rust is selected

    -rust: Optional. Generate Rust code.
           Default: false
           
    -gll: Optional. Generate a GLL parser.
          Default true. False if -knuth or -pager is selected.
                  
    -knuth: Optional. Generate a Knuth LR(1) parser
            Default false

    -pager: Optional. Generate a Pager PGM LR(1) parser.
            Default false

    -resolve_conflicts: Optional. Automatically resolve LR(1) conflicts.
            Default: false. Only used when generating LR(1) parsers.
    
    -bs: Optional. Print BSR statistics (GLL only).
    
    -CPUProf : Optional. Generate a CPU profile. Default false.
        The generated CPU profile is in <cpu.prof>. 
        Use "go tool pprof cpu.prof" to analyse the profile.
```

# Using the generated lexer and parser
1. Create a lexer:  
From an `[]rune`:
```
	lexer.New(input []rune) *Lexer
```
  or from a file. If the file extension us `.md` the lexer will 
  treat all text outside the markdown code blocks as whitespace.
```
	lexer.NewFile(fname string) *Lexer
```
2. Parse the lexer:  
```
	if err, errs := parser.Parse(lex); err != nil {...}
```
3. Check for ambiguities in the parse forest
```
	if bsr.IsAmbiguous() {
		fmt.Println("Error: Ambiguous parse forest")
		bsr.ReportAmbiguous()
		os.Exit(1)
	}
```
Ambiguous BSRs must be resolved by walking the parse forest and ignoring
unwanted children of ambiguous NTs (see [Complete Example](#Complete-Example)).
4. Use the disambiguated parse tree for the further stages of compilation. 
For example, see gogll's [AST builder](ast/build.go).

<a name="Complete-Example"></a>
# Complete Example
The code of following example can be found at [examples/boolx](examples/boolx/boolx.md). 
The example has the following grammar: [boolx.md](examples/boolx/boolx.md), which generates boolean expressions such as: `a | b & c | d & e`:

```
package "github.com/goccmack/gogll/examples/boolx"

Expr :   var
     |   Expr Op Expr
     ;

var : letter ;

Op : "&" | "|" ; 
```
The second alternate above, `Expr : Expr Op Expr`, is ambiguous and can produce an ambiguous parse forest.
The grammar does not enforce operator precedence, 
this has to be done during semantic analysis.

The grammar is compiled by the following command:
```
gogll examples/boolx/boolx.md
```

The test file, [boolx_test.go](examples/boolx/boolx_test.go) shows the steps
required to parse an input string and produce a disambiguated abstract syntax tree:

```
const t1Src = `a | b & c | d & e`

func Test1(t *testing.T) {
```
1. Create a lexer from the input string and parse. Fail if there are parse errors.
```
	if err, errs := parser.Parse(lexer.New([]rune(t1Src))); err != nil {
		fail(errs)
	}

```
2. Build an abstract syntax tree for each root of the parse forest and print them.
```
	for i, r := range bsr.GetRoots() {
		fmt.Printf("%d: %s\n", i, buildExpr(r))
	}
}
```
The input string produces an ambiguous parse forest, which is partially 
disambiguated by applying operator precedence.
We get the following output from this test:
```
> go test -v ./examples/boolx
=== RUN   Test1
0: (a | ((b & c) | (d & e)))
1: <nil>
2: <nil>
3: ((a | (b & c)) | (d & e))
--- PASS: Test1 (0.00s)
PASS
```
The output shows that the parse forest has 4 roots, 2 of which produce valid ASTs 
after disambiguation. The removed trees are syntactically valid by semantically
invalid because they give `|` precedence over `&`. 
Both the remaining ASTs are syntactically and semantically
valid. The AST encodes operator precedence as shown by the parentheses. 
The choice of which valid AST to use for further processing is application specific.

In this example disambiguation by operator precedence is applied during the
AST build. 

Our AST has only one type of node: `Expr`.
```
type ExprType int

const (
	Expr_Var ExprType = iota
	Expr_Expr
)

type Expr struct {
	Type  ExprType
	Var   *token.Token
	Op    *token.Token
	Left  *Expr
	Right *Expr
}

```
A node can represent a variable (`Type` = `Expr_Var`) or an expression (`Type` = `Expr_Expr`).
If the node represents a variable the field `Var` contains the variable token. 
Otherwise `Op` contains the operator token and `Left` and `Right` contain the nodes
of the sub-expressions.

The AST is constructed recursively from each BSR root by the function, `buildExpr`
in [boolx_test.go](examples/boolx/boolx_test.go).

```
/*
Expr :   var
     |   Expr Op Expr
     ;
Op : "&" | "|" ;
*/
func buildExpr(b bsr.BSR) *Expr {
	/*** Expr :   var ***/
	if b.Alternate() == 0 {
		return &Expr{
			Type: Expr_Var,
			Var:  b.GetTChildI(0),
		}
	}

	/*** Expr : Expr Op Expr ***/
	op := b.GetNTChildI(1). // Op is symbol 1 of the Expr rule
				GetTChildI(0) // The operator token is symbol 0 for both alternates of the Op rule

	// Build the left subexpression Node. The subtree for it may be ambiguous.
	left := []*Expr{}
	// b.GetNTChildrenI(0) returns all the valid BSRs for symbol 0 of the body of the rule.
	for _, le := range b.GetNTChildrenI(0) {
		// Add subexpression if it is valid and has precedence over this expression
		if e := buildExpr(le); e != nil && hasPrecedence(e, op) {
			left = append(left, e)
		}
	}
	// No valid subexpressions therefore this whole expression is invalid
	if len(left) == 0 {
		return nil
	}
	// Belts and braces
	if len(left) > 1 {
		panic(fmt.Sprintf("%s has %d left children", b, len(left)))
	}
	// Do the same for the right subexpression

ChangeLog.md
# Change Log

# 2021-02-05 v3.3.0
Add initial support for PEG expressions

# 2020-08-12 v3.2.0
Suppressed tokens added to lexer. See [gogll.md](gogll.md)

# 2020-08-11 v3.1.8
Bug fixed in handling of empty productions.

# 2020-06-28 v3.1.6
This version of gogll supports generation of Rust and Go LR(1) parsers as
well as GLL. Pager's Practical General Method and Knuth's original machine are
both supported.

# v3.1.1
1. Unnecessary match statements removed from generated Rust lexer code.
1. Generated Rust lexer markdown loading performance improved.

# 2020-05-19 v3.1.0
Rust target added.

# 2020-05-13 v3.0.9
1. Global variables removed from parser and BSR.

# 2020-05-01
1. tokens have a method: GetLineColumn()

# 2020-04-30
1. Bugfix in lexer closure algorithm.

# 2020-04-29
1. Some semantic checks implemented
	1. Check that referenced NTs and TokIDs are declared.
1. Gogll stops with an error if the parse forest is ambiguous

# 2020-04-25
1. GoGLL v3.0.3 has both a lexer and parser generated by GoGLL v3.0.2. This version compiles itself.

# 2020-04-??
1. GoGLL v3.0.2 has a GoGLL parser generated by GoGLL v3.0.1 and a hand-written lexer. This is the final bootstrap version.
1. v3.0.2 generates a GoGLL parser with a matching linear-time FSA lexer from one input grammar.

# 2020-04-22
1. GOGLL v3.0.1 features a GoGLL parser generated by GoGLL v3.0.0 and a hand-written
FSA lexer. It is used to bootstrap the GoGLL v3.0.2 which generates both a linear-time FSA 
lexer for a matching GLL parser.

# 2020-04-17 
1. GoGLL v3.0.0 (worktree gogll-lex). This is the basic bootstrap version of GoGLL v3. 
This version is made with Gocc and generates syntax-only GoGLL v3.0.1 parser.

# 2019-09-13
1. Documentation added on how to walk the parse forest. [See](doc/bsr/bsr.md)

## 2019-09-10
1. testSelect is generated as a slice of functions, not a map of functions.
1. follow is generated as separate functions, not a map of functions.

## 2019-09-10 v2.0.0
1. Bug fixed in `ast.NewAnyOf` and `ast.NewNot`
1. Grammar change: `emptyAlt` replace by `empty`. `emptyAlt` was a vestige of 
the bootstrap gocc compiler.

## 2019-09-09
BSR performance improved. 

lexer/lexer.go

// Package lexer is generated by GoGLL. Do not edit.
package lexer

import (
	// "fmt"
	"io/ioutil"
	"strings"
	"unicode"

	"github.com/goccmack/gogll/token"
)

type state int

const nullState state = -1


// Lexer contains both the input slice of runes and the slice of tokens
// parsed from the input
type Lexer struct {
	// I is the input slice of runes
	I      []rune

	// Tokens is the slice of tokens constructed by the lexer from I
	Tokens []*token.Token
}

/*
NewFile constructs a Lexer created from the input file, fname. 

If the input file is a markdown file NewFile process treats all text outside
code blocks as whitespace. All text inside code blocks are treated as input text.

If the input file is a normal text file NewFile treats all text in the inputfile
as input text.
*/
func NewFile(fname string) *Lexer {
	buf, err := ioutil.ReadFile(fname)
	if err != nil {
		panic(err)
	}
	input := []rune(string(buf))
	if strings.HasSuffix(fname, ".md") {
		loadMd(input)
	}
	return New(input)
}

func loadMd(input []rune) {
	i := 0
	text := true
	for i < len(input) {
		if i <= len(input)-3 && input[i] == '`' && input[i+1] == '`' && input[i+2] == '`' {
			text = !text
			for j := 0; j < 3; j++ {
				input[i+j] = ' '
			}
			i += 3
		}
		if i < len(input) {
			if text {
				if input[i] == '\n' {
					input[i] = '\n'
				} else {
					input[i] = ' '
				}
			}
			i += 1
		}
	}
}

/*
New constructs a Lexer from a slice of runes. 

All contents of the input slice are treated as input text.
*/
func New(input []rune) *Lexer {
	lex := &Lexer{
		I:      input,
		Tokens: make([]*token.Token, 0, 2048),
	}
	lext := 0
	for lext < len(lex.I) {
		for lext < len(lex.I) && unicode.IsSpace(lex.I[lext]) {
			lext++
		}
		if lext < len(lex.I) {
			tok := lex.scan(lext)
			lext = tok.Rext()
			if !tok.Suppress() {
				lex.addToken(tok)
			}
		}
	}
	lex.add(token.EOF, len(input), len(input))
	return lex
}

func (l *Lexer) scan(i int) *token.Token {
	// fmt.Printf("lexer.scan\n")
	s, typ, rext := state(0), token.Error, i
	for s != nullState {
		// fmt.Printf("S%d '%c' @ %d\n", s, l.I[rext], rext)
		if rext >= len(l.I) {
			typ = accept[s]
			s = nullState
		} else {
			typ = accept[s]
			s = nextState[s](l.I[rext])
			if s != nullState || typ == token.Error {
				rext++
			}
		}
	}
	return token.New(typ, i, rext, l.I)
}

func escape(r rune) string {
	switch r {
	case '"':
		return "\""
	case '\\':
		return "\\\\"
	case '\r':
		return "\\r"
	case '\n':
		return "\\n"
	case '\t':
		return "\\t"
	}
	return string(r)
}

// GetLineColumn returns the line and column of rune[i] in the input
func (l *Lexer) GetLineColumn(i int) (line, col int) {
	line, col = 1, 1
	for j := 0; j < i; j++ {
		switch l.I[j] {
		case '\n':
			line++
			col = 1
		case '\t':
			col += 4
		default:
			col++
		}
	}
	return
}

// GetLineColumnOfToken returns the line and column of token[i] in the imput
func (l *Lexer) GetLineColumnOfToken(i int) (line, col int) {
	return l.GetLineColumn(l.Tokens[i].Lext())
}

// GetString returns the input string from the left extent of Token[lext] to
// the right extent of Token[rext]
func (l *Lexer) GetString(lext, rext int) string {
	return string(l.I[l.Tokens[lext].Lext():l.Tokens[rext].Rext()])
}

func (l *Lexer) add(t token.Type, lext, rext int) {
	l.addToken(token.New(t, lext, rext, l.I))
}

func (l *Lexer) addToken(tok *token.Token) {
	l.Tokens = append(l.Tokens, tok)
}

func any(r rune, set []rune) bool {
	for _, r1 := range set {
		if r == r1 {
			return true
		}
	}
	return false
}

func not(r rune, set []rune) bool {
	for _, r1 := range set {
		if r == r1 {
			return false
		}
	}
	return true
}

var accept = []token.Type{ 
	token.Error, 
	token.T_0, 
	token.Error, 
	token.Error, 
	token.T_1, 
	token.T_2, 
	token.T_3, 
	token.T_4, 
	token.T_5, 
	token.T_6, 
	token.T_7, 
	token.T_8, 
	token.T_9, 
	token.T_10, 
	token.Error, 
	token.Error, 
	token.Error, 
	token.Error, 
	token.Error, 
	token.Error, 
	token.T_23, 
	token.T_24, 
	token.T_25, 
	token.Error, 
	token.Error, 
	token.T_20, 
	token.Error, 
	token.Error, 
	token.Error, 
	token.T_21, 
	token.T_21, 
	token.T_21, 
	token.T_21, 
	token.T_21, 
	token.T_21, 
	token.T_21, 
	token.T_21, 
	token.T_21, 
	token.T_17, 
	token.T_12, 
	token.T_12, 
	token.T_11, 
	token.T_21, 
	token.T_21, 
	token.T_21, 
	token.T_16, 
	token.T_21, 
	token.T_21, 
	token.T_21, 
	token.T_21, 
	token.T_21, 
	token.T_21, 
	token.T_21, 
	token.T_21, 
	token.T_21, 
	token.T_13, 
	token.T_21, 
	token.T_21, 
	token.T_21, 
	token.T_21, 
	token.T_21, 
	token.T_14, 
	token.T_21, 
	token.T_18, 
	token.T_21, 
	token.T_22, 
	token.T_15, 
	token.T_19, 
}

var nextState = []func(r rune) state{ 
	// Set0
	func(r rune) state {
		switch { 
		case r == '!':
			return 1 
		case r == '"':
			return 2 
		case r == '\'':
			return 3 
		case r == '(':
			return 4 
		case r == ')':
			return 5 
		case r == '.':
			return 6 
		case r == '/':
			return 7 
		case r == ':':
			return 8 
		case r == ';':
			return 9 
		case r == '<':
			return 10 
		case r == '>':
			return 11 
		case r == '[':
			return 12 
		case r == ']':
			return 13 
		case r == 'a':
			return 14 
		case r == 'e':
			return 15 
		case r == 'l':
			return 16 
		case r == 'n':
			return 17 
		case r == 'p':
			return 18 
		case r == 'u':
			return 19 
		case r == '{':
			return 20 
		case r == '|':
			return 21 
		case r == '}':
			return 22 
		case unicode.IsLower(r):
			return 23 
		case unicode.IsUpper(r):
			return 24 
		}
		return nullState
	}, 
	// Set1
	func(r rune) state {
		switch { 
		}
		return nullState
	}, 
	// Set2
	func(r rune) state {
		switch { 
		case r == '"':
			return 25 
		case r == '\\':
			return 26 
		case not(r, []rune{'"','\\'}):
			return 2 
		}
		return nullState
	}, 
	// Set3
	func(r rune) state {
		switch { 
		case r == '\\':
			return 27 
		case not(r, []rune{'\''}):
			return 28 
		}
		return nullState
	}, 
	// Set4
	func(r rune) state {
		switch { 
		}
		return nullState
	}, 
	// Set5
	func(r rune) state {
		switch { 
		}
		return nullState
	}, 
	// Set6
	func(r rune) state {
		switch { 
		}
		return nullState
	}, 
	// Set7
	func(r rune) state {
		switch { 
		}
		return nullState
	}, 
	// Set8
	func(r rune) state {
		switch { 
		}
		return nullState
	}, 
	// Set9
	func(r rune) state {
		switch { 
		}
		return nullState
	}, 
	// Set10
	func(r rune) state {
		switch { 
		}
		return nullState
	}, 
	// Set11
	func(r rune) state {
		switch { 
		}
		return nullState
	}, 
	// Set12
	func(r rune) state {
		switch { 
		}
		return nullState
	}, 
	// Set13
	func(r rune) state {
		switch { 
		}
		return nullState
	}, 
	// Set14
	func(r rune) state {
		switch { 
		case r == '_':
			return 29 
		case r == 'n':
			return 30 
		case unicode.IsLetter(r):
			return 29 
		case unicode.IsNumber(r):
			return 29 
		}
		return nullState
	}, 
	// Set15
	func(r rune) state {
		switch { 
		case r == '_':
			return 29 
		case r == 'm':
			return 31 
		case unicode.IsLetter(r):
			return 29 
		case unicode.IsNumber(r):
			return 29 
		}
		return nullState
	}, 
	// Set16
	func(r rune) state {
		switch { 
		case r == '_':
			return 29 
		case r == 'e':
			return 32 
		case r == 'o':
			return 33 
		case unicode.IsLetter(r):
			return 29 
		case unicode.IsNumber(r):
			return 29 
		}
		return nullState
	}, 
	// Set17
	func(r rune) state {
		switch { 
		case r == '_':
			return 29 
		case r == 'o':
			return 34 
		case r == 'u':
			return 35 
		case unicode.IsLetter(r):
			return 29 
		case unicode.IsNumber(r):
			return 29 
		}
		return nullState
	}, 
	// Set18
	func(r rune) state {
		switch { 
		case r == '_':
			return 29 
		case r == 'a':
			return 36 
		case unicode.IsLetter(r):
			return 29 
		case unicode.IsNumber(r):
			return 29 
		}
		return nullState
	}, 
	// Set19
	func(r rune) state {
		switch { 
		case r == '_':
			return 29 
		case r == 'p':
			return 37 
		case unicode.IsLetter(r):
			return 29 
		case unicode.IsNumber(r):
			return 29 
		}
		return nullState
	}, 
	// Set20
	func(r rune) state {
		switch { 
		}
		return nullState
	}, 
	// Set21
	func(r rune) state {
		switch { 
		}
		return nullState
	}, 
	// Set22
	func(r rune) state {
		switch { 
		}
		return nullState
	}, 
	// Set23
	func(r rune) state {
		switch { 
		case r == '_':
			return 29 
		case unicode.IsLetter(r):
			return 29 
		case unicode.IsNumber(r):
			return 29 
		}
		return nullState
	}, 
	// Set24
	func(r rune) state {
		switch { 
		case r == '_':
			return 38 
		case unicode.IsLetter(r):
			return 38 
		case unicode.IsNumber(r):
			return 38 
		}
		return nullState
	}, 
	// Set25
	func(r rune) state {
		switch { 
		}
		return nullState
	}, 
	// Set26
	func(r rune) state {
		switch { 
		case any(r, []rune{'"','\\','n','r','t'}):
			return 2 
		}
		return nullState
	}, 
	// Set27
	func(r rune) state {
		switch { 
		case any(r, []rune{'\'','\\','n','r','t'}):
			return 39 
		case r == '\'':
			return 39 
		}
		return nullState
	}, 
	// Set28
	func(r rune) state {
		switch { 
		case r == '\'':
			return 40 
		}
		return nullState
	}, 
	// Set29
	func(r rune) state {
		switch { 
		case r == '_':
			return 29 
		case unicode.IsLetter(r):
			return 29 
		case unicode.IsNumber(r):
			return 29 
		}
		return nullState
	}, 
	// Set30
	func(r rune) state {
		switch { 
		case r == '_':
			return 29 
		case r == 'y':
			return 41 
		case unicode.IsLetter(r):
			return 29 
		case unicode.IsNumber(r):
			return 29 
		}
		return nullState
	}, 
	// Set31
	func(r rune) state {
		switch { 
		case r == '_':
			return 29 
		case r == 'p':
			return 42 
		case unicode.IsLetter(r):
			return 29 
		case unicode.IsNumber(r):
			return 29 
		}
		return nullState
	}, 
	// Set32
	func(r rune) state {
		switch { 
		case r == '_':
			return 29 
		case r == 't':
			return 43 
		case unicode.IsLetter(r):
			return 29 
		case unicode.IsNumber(r):
			return 29 
		}
		return nullState
	}, 
	// Set33
	func(r rune) state {
		switch { 
		case r == '_':
			return 29 
		case r == 'w':
			return 44 
		case unicode.IsLetter(r):
			return 29 
		case unicode.IsNumber(r):
			return 29 
		}
		return nullState
	}, 
	// Set34
	func(r rune) state {
		switch { 
		case r == '_':
			return 29 
		case r == 't':
			return 45 
		case unicode.IsLetter(r):
			return 29 
		case unicode.IsNumber(r):
			return 29 
		}
		return nullState
	}, 
	// Set35
	func(r rune) state {
		switch { 
		case r == '_':
			return 29 
		case r == 'm':
			return 46 
		case unicode.IsLetter(r):
			return 29 
		case unicode.IsNumber(r):
			return 29 
		}
		return nullState
	}, 
	// Set36
	func(r rune) state {
		switch { 
		case r == '_':
			return 29 
		case r == 'c':
			return 47 
		case unicode.IsLetter(r):
			return 29 
		case unicode.IsNumber(r):
			return 29 
		}
		return nullState
	}, 
	// Set37
	func(r rune) state {
		switch { 
		case r == '_':
			return 29 
		case r == 'c':
			return 48 
		case unicode.IsLetter(r):
			return 29 
		case unicode.IsNumber(r):
			return 29 
		}
		return nullState
	}, 
	// Set38
	func(r rune) state {
		switch { 
		case r == '_':
			return 38 
		case unicode.IsLetter(r):
			return 38 
		case unicode.IsNumber(r):
			return 38 
		}
		return nullState
	}, 
	// Set39
	func(r rune) state {
		switch { 
		case r == '\'':
			return 40 
		}
		return nullState
	}, 
	// Set40
	func(r rune) state {
		switch { 
		}
		return nullState
	}, 
	// Set41
	func(r rune) state {
		switch { 
		case r == '_':
			return 29 
		case unicode.IsLetter(r):
			return 29 
		case unicode.IsNumber(r):
			return 29 
		}
		return nullState
	}, 
	// Set42
	func(r rune) state {
		switch { 
		case r == '_':
			return 29 
		case r == 't':
			return 49 
		case unicode.IsLetter(r):
			return 29 
		case unicode.IsNumber(r):
			return 29 
		}
		return nullState
	}, 
	// Set43
	func(r rune) state {
		switch { 
		case r == '_':
			return 29 
		case r == 't':
			return 50 
		case unicode.IsLetter(r):
			return 29 
		case unicode.IsNumber(r):
			return 29 
		}
		return nullState
	}, 
	// Set44
	func(r rune) state {
		switch { 
		case r == '_':
			return 29 
		case r == 'c':
			return 51 
		case unicode.IsLetter(r):
			return 29 
		case unicode.IsNumber(r):
			return 29 
		}
		return nullState
	}, 
	// Set45
	func(r rune) state {
		switch { 
		case r == '_':
			return 29 
		case unicode.IsLetter(r):
			return 29 
		case unicode.IsNumber(r):
			return 29 
		}
		return nullState
	}, 
	// Set46
	func(r rune) state {
		switch { 
		case r == '_':
			return 29 
		case r == 'b':
			return 52 
		case unicode.IsLetter(r):
			return 29 
		case unicode.IsNumber(r):
			return 29 
		}
		return nullState
	}, 
	// Set47
	func(r rune) state {
		switch { 
		case r == '_':
			return 29 
		case r == 'k':
			return 53 
		case unicode.IsLetter(r):
			return 29 
		case unicode.IsNumber(r):
			return 29 
		}
		return nullState
	}, 
	// Set48
	func(r rune) state {
		switch { 
		case r == '_':
			return 29 
		case r == 'a':
			return 54 
		case unicode.IsLetter(r):
			return 29 
		case unicode.IsNumber(r):
			return 29 
		}
		return nullState
	}, 
	// Set49
	func(r rune) state {
		switch { 
		case r == '_':
			return 29 
		case r == 'y':
			return 55 
		case unicode.IsLetter(r):
			return 29 
		case unicode.IsNumber(r):
			return 29 
		}
		return nullState
	}, 
	// Set50
	func(r rune) state {
		switch { 
		case r == '_':
			return 29 
		case r == 'e':
			return 56 
		case unicode.IsLetter(r):
			return 29 
		case unicode.IsNumber(r):
			return 29 
		}
		return nullState
	}, 
	// Set51
	func(r rune) state {
		switch { 
		case r == '_':
			return 29 
		case r == 'a':
			return 57 
		case unicode.IsLetter(r):
			return 29 
		case unicode.IsNumber(r):
			return 29 
		}
		return nullState
	}, 
	// Set52
	func(r rune) state {
		switch { 
		case r == '_':
			return 29 
		case r == 'e':
			return 58 
		case unicode.IsLetter(r):
			return 29 
		case unicode.IsNumber(r):
			return 29 
		}
		return nullState
	}, 
	// Set53
	func(r rune) state {
		switch { 
		case r == '_':
			return 29 
		case r == 'a':
			return 59 
		case unicode.IsLetter(r):
			return 29 
		case unicode.IsNumber(r):
			return 29 
		}
		return nullState
	}, 
	// Set54
	func(r rune) state {
		switch { 
		case r == '_':
			return 29 
		case r == 's':
			return 60 
		case unicode.IsLetter(r):
			return 29 
		case unicode.IsNumber(r):
			return 29 
		}
		return nullState
	}, 
	// Set55
	func(r rune) state {
		switch { 
		case r == '_':
			return 29 
		case unicode.IsLetter(r):
			return 29 
		case unicode.IsNumber(r):
			return 29 
		}
		return nullState
	}, 
	// Set56
	func(r rune) state {
		switch { 
		case r == '_':
			return 29 
		case r == 'r':
			return 61 
		case unicode.IsLetter(r):
			return 29 
		case unicode.IsNumber(r):
			return 29 
		}
		return nullState
	}, 
	// Set57
	func(r rune) state {
		switch { 
		case r == '_':
			return 29 
		case r == 's':
			return 62 
		case unicode.IsLetter(r):
			return 29 
		case unicode.IsNumber(r):
			return 29 
		}
		return nullState
	}, 
	// Set58
	func(r rune) state {
		switch { 
		case r == '_':
			return 29 
		case r == 'r':
			return 63 
		case unicode.IsLetter(r):
			return 29 
		case unicode.IsNumber(r):
			return 29 
		}
		return nullState
	}, 
	// Set59
	func(r rune) state {
		switch { 
		case r == '_':
			return 29 
		case r == 'g':
			return 64 
		case unicode.IsLetter(r):
			return 29 
		case unicode.IsNumber(r):
			return 29 
		}
		return nullState
	}, 
	// Set60
	func(r rune) state {
		switch { 
		case r == '_':
			return 29 
		case r == 'e':
			return 65 
		case unicode.IsLetter(r):
			return 29 
		case unicode.IsNumber(r):
			return 29 
		}
		return nullState
	}, 
	// Set61
	func(r rune) state {
		switch { 
		case r == '_':
			return 29 
		case unicode.IsLetter(r):
			return 29 
		case unicode.IsNumber(r):
			return 29 
		}
		return nullState
	}, 
	// Set62
	func(r rune) state {
		switch { 
		case r == '_':
			return 29 
		case r == 'e':
			return 66 
		case unicode.IsLetter(r):
			return 29 
		case unicode.IsNumber(r):
			return 29 
		}
		return nullState
	}, 
	// Set63
	func(r rune) state {
		switch { 
		case r == '_':
			return 29 
		case unicode.IsLetter(r):
			return 29 
		case unicode.IsNumber(r):
			return 29 
		}
		return nullState
	}, 
	// Set64
	func(r rune) state {
		switch { 
		case r == '_':
			return 29 
		case r == 'e':
			return 67 
		case unicode.IsLetter(r):
			return 29 
		case unicode.IsNumber(r):
			return 29 
		}
		return nullState
	}, 
	// Set65
	func(r rune) state {
		switch { 
		case r == '_':
			return 29 
		case unicode.IsLetter(r):
			return 29 
		case unicode.IsNumber(r):
			return 29 
		}
		return nullState
	}, 
	// Set66
	func(r rune) state {
		switch { 
		case r == '_':
			return 29 
		case unicode.IsLetter(r):
			return 29 
		case unicode.IsNumber(r):
			return 29 
		}
		return nullState
	}, 
	// Set67
	func(r rune) state {
		switch { 
		case r == '_':
			return 29 
		case unicode.IsLetter(r):
			return 29 
		case unicode.IsNumber(r):
			return 29 
		}
		return nullState
	}, 
}
